### PR TITLE
DM-39900: Develop improved metrics for completeness based on injected sources

### PR DESCRIPTION
**Pull request by BrunoSanchez**
Adding metrics that calculate fraction of fakes found.

Next iteration is to provide a completeness per bandpass in magnitude space.

### CODE REVIEW

File: pipelines/apFakesMetrics.yaml
Comment (isullivan) [inline pipelines/apFakesMetrics.yaml:1]: This should probably be named `apDetectorVisitFakesCore.yaml`

File: pipelines/apFakesMetrics.yaml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      # From 6 to infinity
+      atools.numFoundFakesDiaSNR6tInfMetric: FractionFoundFakesDiaSnrMetric
+      atools.numFoundFakesDiaSNR6tInfMetric.snrMin: 6
+      connections.outputName: diaFakesDetectorVisitCore
+      python: |
+        from lsst.analysis.tools.atools import FractionFoundFakesDiaSnrMetric
+  analyzeDiaFakesVisitCore:
+    class: lsst.analysis.tools.tasks.DiaFakesVisitAnalysisTask
+    config:
+      # connections.coaddName: parameters.coaddName
```
Comment (isullivan) [inline pipelines/apFakesMetrics.yaml:41]: Remove commented-out line

File: python/lsst/analysis/tools/tasks/diaFakesDetectorVisitAnalysis.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+from lsst.pipe.base import connectionTypes as ct
+
+from ..interfaces import AnalysisBaseConfig, AnalysisBaseConnections, AnalysisPipelineTask
+
+
+class DiaFakesDetectorVisitAnalysisConnections(
+    AnalysisBaseConnections,
+    dimensions=("visit", "band", "detector"),
```
[the remark this answers is not in the document]
Comment (isullivan) [reply python/lsst/analysis/tools/tasks/diaFakesDetectorVisitAnalysis.py:36]: Probably should be `("instrument", "visit", "detector")`

File: python/lsst/analysis/tools/tasks/diaFakesDetectorVisitAnalysis.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    defaultTemplates={"coaddName": "goodSeeing", "fakesType": "fakes_"},
+):
+    data = ct.Input(
+        doc="CcdVisit-based Matched fake to load from the butler",
+        name="{fakesType}{coaddName}Diff_matchDiaSrc",
+        storageClass="DataFrame",
+        deferLoad=True,
```
Comment (isullivan) [inline python/lsst/analysis/tools/tasks/diaFakesDetectorVisitAnalysis.py:43]: Why `deferLoad`?

File: python/lsst/analysis/tools/tasks/diaFakesVisitAnalysis.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+class DiaFakesVisitAnalysisConnections(
+    AnalysisBaseConnections,
+    dimensions=("visit", "band"),
+    defaultTemplates={"coaddName": "goodSeeing", "fakesType": "fakes_"},
+):
+    data = ct.Input(
+        doc="CcdVisit-based Matched fake to load from the butler",
+        name="{fakesType}{coaddName}Diff_matchDiaSourceTable",
+        storageClass="DataFrame",
+        deferLoad=True,
```
Comment (isullivan) [inline python/lsst/analysis/tools/tasks/diaFakesVisitAnalysis.py:43]: As above, why `deferLoad`?

File: python/lsst/analysis/tools/tasks/diaFakesVisitAnalysis.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = (
+    "DiaFakesVisitAnalysisConnections",
+    "DiaFakesVisitAnalysisConfig",
+    "DiaFakesVisitAnalysisTask",
+)
+
+from lsst.pipe.base import connectionTypes as ct
```
Comment (isullivan) [inline python/lsst/analysis/tools/tasks/diaFakesVisitAnalysis.py:29]: As above, don't import as `ct`

File: python/lsst/analysis/tools/tasks/diaFakesDetectorVisitAnalysis.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = (
+    "DiaFakesDetectorVisitAnalysisConnections",
```
Comment (isullivan) [inline python/lsst/analysis/tools/tasks/diaFakesDetectorVisitAnalysis.py:24]: I don't think the connections class needs to be in `__all__`

File: python/lsst/analysis/tools/tasks/diaFakesVisitAnalysis.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = (
+    "DiaFakesVisitAnalysisConnections",
```
Comment (isullivan) [inline python/lsst/analysis/tools/tasks/diaFakesVisitAnalysis.py:24]: I don't think the connections class needs to be in `__all__`

**Review by isullivan — COMMENTED** [review]
I ran out of time to finish the review today, but I will come back to it tomorrow. I think I am mostly done, but I want to take another look before signing off.

**Review by isullivan — APPROVED** [review]
Looks good! I had a few cleanup and naming comments.

File: python/lsst/analysis/tools/tasks/diaFakesDetectorVisitAnalysis.py (hunk 3 above)
Comment (BrunoSanchez) [reply python/lsst/analysis/tools/tasks/diaFakesDetectorVisitAnalysis.py:43]: Not sure why this is needed. But my test runs conclude that removing this makes the pipeline fail. Maybe it's a good topic for standup.

**Pull request merged**